kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/GCSToBQLoadRunnable.java

```diff
@@ -0,0 +1,232 @@
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * A Runnable that runs a GCS to BQ Load task.
+ *
+ * <p>This task goes through the given GCS bucket, and takes as many blobs as a single load job per
+ * table can handle (as defined here: https://cloud.google.com/bigquery/quotas#load_jobs) and runs
+ * those load jobs. Blobs are deleted (only) once a load job involving that blob succeeds.
+ */
+public class GCSToBQLoadRunnable implements Runnable {
+  private static final Logger logger = LoggerFactory.getLogger(GCSToBQLoadRunnable.class);
+
+  private final BigQuery bigQuery;
+  private final Bucket bucket;
+  private final Map<Job, List<Blob>> activeJobs;
+  private final Set<Blob> claimedBlobs;
+
+  // these numbers are intended to try to make this task not excede Google Cloud Quotas.
+  // see: https://cloud.google.com/bigquery/quotas#load_jobs
+  private static int FILE_LOAD_LIMIT = 10000;
+  // max number of files we can load in a single load job
+  private static long MAX_LOAD_SIZE_B = 15 * 1000000000000L; // 15TB
+  // max total size (in bytes) of the files we can load in a single load job.
+  private static String SOURCE_URI_FORMAT = "gs://%s/%s";
+  public static final Pattern METADATA_TABLE_PATTERN =
+          Pattern.compile("(?<project>[^:]+):(?<dataset>[^.]+).(?<table>.+)");
+
+  /**
+   * Create a {@link GCSToBQLoadRunnable} with the given bigquery, bucket, and ms wait interval.
+   * @param bigQuery the {@link BigQuery} instance.
+   * @param bucket the the GCS bucket to read from.
+   */
+  public GCSToBQLoadRunnable(BigQuery bigQuery, Bucket bucket) {
+    this.bigQuery = bigQuery;
+    this.bucket = bucket;
+    this.activeJobs = new HashMap<>();
+    this.claimedBlobs = new HashSet<>();
+  }
+
+  private Map<TableId, List<Blob>> getBlobsUpToLimit() {
+    Map<TableId, List<Blob>> tableToURIs = new HashMap<>();
+    //List<String> URIs = new ArrayList<>();
+    Map<TableId, Long> tableToCurrentLoadSize = new HashMap<>();
+
+    Page<Blob> list = bucket.list();
+    for (Blob blob : list.iterateAll()) {
+      TableId table = getTableFromBlob(blob);
+
+      if (table == null || claimedBlobs.contains(blob)) {
+        // don't do anything if:
+        // 1. we don't know what table this should be uploaded to or
+        // 2. this blob is already claimed by a currently-running job.
+        continue;
+      }
+
+      if (!tableToURIs.containsKey(table)) {
+        // initialize maps, if we haven't seen this table before.
+        tableToURIs.put(table, new ArrayList<>());
+        tableToCurrentLoadSize.put(table, 0L);
+      }
+
+      long newSize = tableToCurrentLoadSize.get(table) + blob.getSize();
+      // if this file does not cause us to exceed our per-request quota limits...
+      if (newSize <= MAX_LOAD_SIZE_B && tableToURIs.get(table).size() < FILE_LOAD_LIMIT) {
```

Comment:
Paranoia: not sure if <= or < is appropriate, so I vote for < when cehcking MAX_LOAD_SIZE_B

Reply:
fair enough.